### ISSUE TITLE
Use generic type parameters names in signatures

### ISSUE THREAD

**Issue by atifaziz** [lijunle/Vsxmd#30]
Signatures of generic methods are not very readable at the moment, like [<code>TestGenericType``2</code>](https://github.com/lijunle/Vsxmd/blob/50093dd10dc38a80280b6b5fa4fc550cd2f4574b/Vsxmd/Vsxmd.md#nthlast1sourceindex-method--). Would be better to replace ``2 with the generic type parameter names `T1` and `T2`, especially when they can be pulled out of `<typeparam>`.

**Comment by lijunle** [comment]
Hmm... that is generic hell...

If we are going to extract `T1` and `T2` to ` ``2 ` in the caption, how about their [`<see>` reference](https://github.com/lijunle/Vsxmd/blob/50093dd10dc38a80280b6b5fa4fc550cd2f4574b/Vsxmd/Vsxmd.md#M-Vsxmd-Program-Test-TestGenericRefence)? How about the generic type in [the parameter type](https://github.com/lijunle/Vsxmd/blob/50093dd10dc38a80280b6b5fa4fc550cd2f4574b/Vsxmd/Vsxmd.md#M-Vsxmd-Program-Test-TestGenericParameter%60%602-System-Linq-Expressions-Expression%7BSystem-Func%7B%60%600,%60%601,System-String%7D%7D-)? So, for simplify, I keep whatever I get from XML and write them on Markdown.

Basically, if fully support generic type, the current design needs to update. It needs at least two passes - first to build type and information mapping, second to translate to Markdown.

Ideas?

**Comment by atifaziz** [comment]
Yes, ideally it would apply to wherever a generic type can appear. The trouble is that while the documentation looks great for simple types & members, it's sub-optimal for code heavy on generics. See the [generated doc](https://gist.github.com/atifaziz/ab6ed0000ee1f33cbea4) for [MoreLINQ](https://www.nuget.org/packages/morelinq/) as an example.

> Basically, if fully support generic type, the current design needs to update. It needs at least two passes - first to build type and information mapping, second to translate to Markdown.

I'm afraid it would come down to something like that because you might even need information not available in the XML doc file, like visibility of the type. That's because the generated docs should only report on the public API by default (something I was going open as a separate issue but may need to be part of the same design). The generator could try and use reflection if the DLL is available to get richer type information but it does need some tricky work to map between the type & signature encoding in the XML doc attributes & reflection API queries.

**Comment by lijunle** [comment]
Yes, I agree that the XML information is not enough to build a great document - we need to reflect the DLL. That could be a big project and need more time on it. Do you have any ideas and design it?

BTW, please open the issue about the publish API. I want to keep it open, at least. :)

**Comment by atifaziz** [comment]
> BTW, please open the issue about the publish API

Thanks for [doing this](https://github.com/lijunle/Vsxmd/issues/31).

>  That could be a big project and need more time on it.

Yes, I was afraid of that and wasn't sure what's your road map and scope with this project but without it, I feel that Vsxmd is just short of being absolutely the next great thing since sliced bread. :wink: 

> Do you have any ideas and design it?

OK, here go my 2¢ cents…Perhaps this can be done as a pre-processor that's a completely separate tool and which can be managed separately (perhaps even as a separate GitHub project). This tool would be responsible for reflecting an assembly & enhancing its XML doc with extra information using new XML elements from a custom namespace. Another option would be to generate a *buddy* XML with metadata. Either way, Vsxmd could then use that information to generate a richer doc & fallback otherwise to its current implementation. With this approach, you have two passes but as separate tools, each of which can enhance separately. The pre-processor could potentially benefit other doc generators who would have to solve the same problem.

**Comment by lijunle** [comment]
Thanks for your kind word.

I love the idea to make a preprocessor to do the reflection and other dirty work. Then, combine two tools to generate the final document. I prefer the enhanced XML doc way (method 1).

Personally, I would like to kick off a F# project for learning. Do you want to join?

**Comment by atifaziz** [comment]
@lijunle Sorry, got busy. Meanwhile, perhaps a benefit of this delay has been that the [DocFX project has gone open source](https://github.com/dotnet/docfx). I was wondering about two things:

1. How do you see the roadmap and future of Vsxmd with respect to DocFX.
2. Isn't the [DocFX metadata](http://dotnet.github.io/docfx/spec/metadata_format_spec.html) solving the same problem as the preprocessor we might be talking about here? Could Vsxmd use it?

I haven't been able to take DocFX for a complete spin yet as [I had trouble building it](https://github.com/dotnet/docfx/issues/151).

**Comment by lijunle** [comment]
@atifaziz Sorry that reply you after a long time. I am quite busy and work on another project on my spare time, not having much time to touch this project's code.

I ever follow the guide of DocFX. It finally build up a website with API documents and tutorial documents. It did a good job that, only expose the public APIs. So, for your second question, the answer could be yes, although I don't know how to leverage the DocFX API.

In regards to your question 1, from their tutorial, the final result is serving a static website (HTML files) from my local machine. The API output is a JSON file, if I remember right. It means none of the output can be viewed directly in Github.

The purpose of Vsxmd is, make some easy setup, generate one or more markdown files, serve them in Github (or GitLab, etc.). No complex setup, no separated project, no document server.

For the roadmap, here are two points I am considering.

- Only expose the public API document. (Hope to leverage DocFX)
- Fail the CI build if the generated document is different from the checked-in one. This enforces each modification to public API being reviewed.

**Comment by atifaziz** [comment]
> The API output is a JSON file, if I remember right. It means none of the output can be viewed directly in Github.

Are you sure? Because that would be a real shame? Perhaps the JSON is just an intermediate file? In any case, I never managed to dig into it as I never got to [build it properly until just now](https://github.com/dotnet/docfx/issues/151#issuecomment-188985849). 

I agree that the charm of Vsxmd could be that it's small & simple. DocFX looks big & complex but it would be great to be able to leverage the metadata from DocFX. May be it would be good to paint the scenario to the DocFX team and get their input on what's the best way to connect the tools & APIs.

**Comment by lijunle** [comment]
Some finds after playing DocFX:

## DocFX.MSBuild

The [doxfx.msbuild](https://www.nuget.org/packages/docfx.msbuild/) package is awesome. It prepare the default configurations and generate the HTML site in one click.

However, its drawback is only supporting the default theme. Besides, the user still need to handle the awful DocFX configuration.

## DocFX Theme

The DocFX [theme](https://github.com/dotnet/docfx/tree/master/src/docfx.website.themes) is excellent. It support exporting as HTML, Markdown and even [Angular SPA](https://github.com/dotnet/docfx/tree/master/src/docfx.website.themes/angular).

But, the DocFX has a [bad documentation](http://dotnet.github.io/docfx/tutorial/howto_create_custom_template.html). :cry: From my attempt, only the default theme is working, other themes in source code need a small tricky hack to work.

Currently, there is no Markdown theme. *This project can create a Markdown theme and some helpers to simplify the configuration.*

## Metadata

The DocFX generated metadata is YAML format and has awesome information. But DocFX does not expose an API for it. The nearest entry point is [ExtractMetadataWorker.ExtractMetadataAsync](https://github.com/dotnet/docfx/blob/c967bd09f51c55944a3752cae893244544e0edb8/src/Microsoft.DocAsCode.EntityModel/ExtractMetadata/ExtractMetadataWorker.cs#L60). This method needs complex configuration as input and store the result as disk files.

However, this file is refactored to another project and location. I am not sure the final structure.

---

## Attachment: Vsxmd.Converter metadata

```yaml
items:
- uid: Vsxmd.IConverter
  id: IConverter
  parent: Vsxmd
  children:
  - Vsxmd.IConverter.ToMarkdown
  langs:
  - csharp
  - vb
  name: IConverter
  fullName: Vsxmd.IConverter
  type: Interface
  source:
    remote:
      path: Vsxmd/IConverter.cs
      branch: update-deps
      repo: git@github.com:lijunle/Vsxmd.git
    id: IConverter
    path: Vsxmd/IConverter.cs
    startLine: 11
  assemblies:
  - Vsxmd
  namespace: Vsxmd
  summary: "\r\nConverter for XML docuement to Markdown syntax conversion.\r\n"
  example: []
  syntax:
    content: public interface IConverter
    content.vb: Public Interface IConverter
  modifiers.csharp:
  - public
  - interface
  modifiers.vb:
  - Public
  - Interface
- uid: Vsxmd.IConverter.ToMarkdown
  id: ToMarkdown
  parent: Vsxmd.IConverter
  langs:
  - csharp
  - vb
  name: ToMarkdown()
  fullName: Vsxmd.IConverter.ToMarkdown()
  type: Method
  source:
    remote:
      path: Vsxmd/IConverter.cs
      branch: update-deps
      repo: git@github.com:lijunle/Vsxmd.git
    id: ToMarkdown
    path: Vsxmd/IConverter.cs
    startLine: 17
  assemblies:
  - Vsxmd
  namespace: Vsxmd
  summary: "\r\nConvert to Markdown syntax.\r\n"
  example: []
  syntax:
    content: string ToMarkdown()
    content.vb: Function ToMarkdown As String
    return:
      type: System.String
      description: The generated Markdown content.
references:
- uid: Vsxmd
  isExternal: false
  name: Vsxmd
  fullName: Vsxmd
- uid: System.String
  isExternal: true
  name: String
  fullName: System.String
```

**Comment by lijunle** [comment]
Time past but I don't think anybody is going to implement this. I am closing this issue. If there are more questions, please reply.
